README.md
# CardScan
This repository contains the open source code for [CardScan](https://www.getbouncer.com)'s react-native implementation.

[CardScan](https://www.getbouncer.com/) is a relatively small library that provides fast and accurate payment card scanning.

CardScan is the foundation for CardVerify enterprise libraries, which validate the authenticity of payment cards as they are scanned.

![Android Lint](https://github.com/getbouncer/react-native-cardscan/workflows/Android%20Lint/badge.svg)
![Android Unit Tests](https://github.com/getbouncer/react-native-cardscan/workflows/Android%20Unit%20Tests/badge.svg)
![Android Instrumentation Tests](https://github.com/getbouncer/react-native-cardscan/workflows/Android%20Instrumentation%20Tests/badge.svg)
![Release](https://github.com/getbouncer/react-native-cardscan/workflows/Release/badge.svg)
[![GitHub release](https://img.shields.io/github/release/getbouncer/react-native-cardscan.svg?maxAge=60)](https://github.com/getbouncer/react-native-cardscan/releases)

Native libraries for [android](https://github.com/getbouncer/cardscan-andorid) and [iOS](https://github.com/getbouncer/cardscan-ios) are also available in github.

CardScan is open source, and available under a free-to-try license. See the [license](#license) section for details.

![demo](docs/images/demo.gif)

## Contents
* [Requirements](#requirements)
* [Demo](#demo)
* [Integration](#integration)
* [Customizing](#customizing)
* [Developing](#developing)
* [Troubleshooting](#troubleshooting)
* [Authors](#authors)
* [License](#license)

## Requirements
* Android API level 21 or higher
* iOS version 11 or higher

## Demo
This repository contains a demonstration app for the CardScan product. To build and run the demo app, follow the instructions in the [example app documentation](https://docs.getbouncer.com/card-scan/react-native-integration-guide#example-app).

## Integration
See the [integration documentation](https://docs.getbouncer.com/card-scan/react-native-integration-guide) in the Bouncer Docs.

### Provisioning an API key
CardScan requires a valid API key to run. To provision an API key, visit the [Bouncer API console](https://api.getbouncer.com/console).

### Name and expiration extraction support (BETA)
To test name and/or expiration extraction, please first provision an API key, then reach out to [support@getbouncer.com](mailto:support@getbouncer.com) with details about your use case and estimated volumes.

Follow the [configuration guide](https://docs.getbouncer.com/card-scan/react-native-integration-guide#configuration) to enable name and expiry extraction support.

## Troubleshooting
See the [troubleshooting page](https://docs.getbouncer.com/card-scan/react-native-integration-guide/troubleshooting) in the Bouncer Docs to check if we have covered common issues.

## Authors
Adam Wushensky, Jaime Park, Sam King, and Zain ul Abi Din

## License
This library is available under paid and free licenses. See the [LICENSE](LICENSE) file for the full license text.

### Quick summary
In short, this library will remain free forever for non-commercial applications, but use by commercial applications is limited to 90 days, after which time a licensing agreement is required. We're also adding some legal liability protections.

After this period commercial applications need to convert to a licensing agreement to continue to use this library.
* Details of licensing (pricing, etc) are available at [https://getbouncer.com/pricing](https://getbouncer.com/pricing), or you can contact us at [license@getbouncer.com](mailto:license@getbouncer.com).

### More detailed summary
What's allowed under the license:
* Free use for any app for 90 days (for demos, evaluations, hackathons, etc).
* Contributions (contributors must agree to the [Contributor License Agreement](Contributor%20License%20Agreement))
* Any modifications as needed to work in your app

What's not allowed under the license:
* Commercial applications using the license for longer than 90 days without a license agreement. 
* Using us now in a commercial app today? No worries! Just email [license@getbouncer.com](mailto:license@getbouncer.com) and we’ll get you set up.
* Redistribution under a different license
* Removing attribution
* Modifying logos
* Indemnification: using this free software is ‘at your own risk’, so you can’t sue Bouncer Technologies, Inc. for problems caused by this library

Questions? Concerns? Please email us at [license@getbouncer.com](mailto:license@getbouncer.com) or ask us on [slack](https://getbouncer.slack.com).


CHANGELOG.md
# Changelog

## [2.0.0](https://github.com/getbouncer/react-native-cardscan/tree/2.0.0) (2021-10-04)

**Closed issues:**

- Getting version error during pod install [\#153](https://github.com/getbouncer/react-native-cardscan/issues/153)
- Value of card expiry month, year [\#150](https://github.com/getbouncer/react-native-cardscan/issues/150)
- Trying to install package in react native causes build failure [\#129](https://github.com/getbouncer/react-native-cardscan/issues/129)
- \[Customizing\] How to customize the câmera overlay? [\#120](https://github.com/getbouncer/react-native-cardscan/issues/120)
- Credit Card Image [\#116](https://github.com/getbouncer/react-native-cardscan/issues/116)
- iOS: not building project issue \(ld: symbol\(s\) not found for architecture arm64\) [\#100](https://github.com/getbouncer/react-native-cardscan/issues/100)
- \> Task :react-native-cardscan:packageDebugAssets FAILED [\#95](https://github.com/getbouncer/react-native-cardscan/issues/95)
- iOS Build Issue [\#56](https://github.com/getbouncer/react-native-cardscan/issues/56)

**Merged pull requests:**

- Release 2.0.0 [\#155](https://github.com/getbouncer/react-native-cardscan/pull/155) ([jaimepark-stripe](https://github.com/jaimepark-stripe))
- Create RN UI Customization interface [\#152](https://github.com/getbouncer/react-native-cardscan/pull/152) ([jaimepark-stripe](https://github.com/jaimepark-stripe))
- Update react dependency from React to React-Core [\#149](https://github.com/getbouncer/react-native-cardscan/pull/149) ([kingst-stripe](https://github.com/kingst-stripe))
- Clean up the example app [\#148](https://github.com/getbouncer/react-native-cardscan/pull/148) ([kingst-stripe](https://github.com/kingst-stripe))
- Update CODEOWNERS [\#146](https://github.com/getbouncer/react-native-cardscan/pull/146) ([jaimepark-stripe](https://github.com/jaimepark-stripe))

## [1.1.30](https://github.com/getbouncer/react-native-cardscan/tree/1.1.30) (2021-04-22)

## [1.1.29](https://github.com/getbouncer/react-native-cardscan/tree/1.1.29) (2021-04-16)

**Closed issues:**

- Is it possible to scan cardholderName on android? [\#106](https://github.com/getbouncer/react-native-cardscan/issues/106)

**Merged pull requests:**

- Update android version [\#115](https://github.com/getbouncer/react-native-cardscan/pull/115) ([awushensky](https://github.com/awushensky))
- Bump to CardScan 2.0.9 [\#103](https://github.com/getbouncer/react-native-cardscan/pull/103) ([jaimejiyepark](https://github.com/jaimejiyepark))

## [1.1.28](https://github.com/getbouncer/react-native-cardscan/tree/1.1.28) (2021-03-10)

**Closed issues:**

- CVC number not scanned [\#84](https://github.com/getbouncer/react-native-cardscan/issues/84)

**Merged pull requests:**

- Upgrade android version [\#117](https://github.com/getbouncer/react-native-cardscan/pull/117) ([awushensky](https://github.com/awushensky))

## [1.1.26](https://github.com/getbouncer/react-native-cardscan/tree/1.1.26) (2020-12-04)

**Merged pull requests:**

- Upgrade native libraries [\#83](https://github.com/getbouncer/react-native-cardscan/pull/83) ([awushensky](https://github.com/awushensky))
- Update native libraries [\#72](https://github.com/getbouncer/react-native-cardscan/pull/72) ([awushensky](https://github.com/awushensky))
- Fix NPE [\#70](https://github.com/getbouncer/react-native-cardscan/pull/70) ([awushensky](https://github.com/awushensky))
- Upgrade android [\#69](https://github.com/getbouncer/react-native-cardscan/pull/69) ([awushensky](https://github.com/awushensky))
- Upgrade cardscan [\#67](https://github.com/getbouncer/react-native-cardscan/pull/67) ([awushensky](https://github.com/awushensky))
- Upgrade native cardscan [\#65](https://github.com/getbouncer/react-native-cardscan/pull/65) ([awushensky](https://github.com/awushensky))
- Upgrade cardscan [\#64](https://github.com/getbouncer/react-native-cardscan/pull/64) ([awushensky](https://github.com/awushensky))

## [1.1.16](https://github.com/getbouncer/react-native-cardscan/tree/1.1.16) (2020-10-06)

**Merged pull requests:**

- Upgrade android [\#62](https://github.com/getbouncer/react-native-cardscan/pull/62) ([awushensky](https://github.com/awushensky))

## [1.1.15](https://github.com/getbouncer/react-native-cardscan/tree/1.1.15) (2020-10-05)

**Merged pull requests:**

- Pin cardscan version [\#61](https://github.com/getbouncer/react-native-cardscan/pull/61) ([awushensky](https://github.com/awushensky))
- Update README.md [\#60](https://github.com/getbouncer/react-native-cardscan/pull/60) ([jaimejiyepark](https://github.com/jaimejiyepark))

## [1.1.14](https://github.com/getbouncer/react-native-cardscan/tree/1.1.14) (2020-10-02)

**Merged pull requests:**

- Update android version [\#59](https://github.com/getbouncer/react-native-cardscan/pull/59) ([awushensky](https://github.com/awushensky))

## [1.1.12](https://github.com/getbouncer/react-native-cardscan/tree/1.1.12) (2020-10-02)

**Merged pull requests:**

- Update android version [\#58](https://github.com/getbouncer/react-native-cardscan/pull/58) ([awushensky](https://github.com/awushensky))

## [1.1.11](https://github.com/getbouncer/react-native-cardscan/tree/1.1.11) (2020-09-23)

**Merged pull requests:**

- Bump cardscan version [\#52](https://github.com/getbouncer/react-native-cardscan/pull/52) ([awushensky](https://github.com/awushensky))
- Fix dismiss view [\#51](https://github.com/getbouncer/react-native-cardscan/pull/51) ([jaimejiyepark](https://github.com/jaimejiyepark))

## [1.1.10](https://github.com/getbouncer/react-native-cardscan/tree/1.1.10) (2020-09-18)

**Merged pull requests:**

- Fix package lock [\#50](https://github.com/getbouncer/react-native-cardscan/pull/50) ([awushensky](https://github.com/awushensky))
- Upgrade android [\#49](https://github.com/getbouncer/react-native-cardscan/pull/49) ([awushensky](https://github.com/awushensky))

ios/RNCardScan.swift
//
//  RNCardScan.swift
//  RNCardScan
//
//  Created by Jaime Park on 9/22/21.
//  Copyright © 2021 Facebook. All rights reserved.
//

import CardScan
import Foundation
import UIKit

@available(iOS 11.2, *)
@objc(RNCardscan)
class RNCardScan: NSObject {
    var resolve: RCTPromiseResolveBlock?
    var styleDictionary: NSDictionary?

    override init() {
        super.init()
    }

    @objc class func requiresMainQueueSetup() -> Bool {
        return true
    }

    @objc func isSupportedAsync(
        _ resolve: RCTPromiseResolveBlock,
        _ reject: RCTPromiseRejectBlock
    ) -> Void {
        if true {
            resolve([SimpleScanViewController.isCompatible()])
        } else {
            reject(nil, nil, nil)
        }
    }

    @objc func setiOSScanViewStyle(_ styleDictionary: NSDictionary) {
        self.styleDictionary = styleDictionary
    }

    @objc func scan(
        _ resolve: @escaping RCTPromiseResolveBlock,
        _ reject: RCTPromiseRejectBlock
    ) -> Void {
      self.resolve = resolve

      DispatchQueue.main.async {
        let topViewController = self.getTopViewController()
        let vc = RNCardScanViewController(viewStyle: self.styleDictionary)
        vc.delegate = self
        topViewController?.present(vc, animated: true, completion: nil)
      }
    }

    func getTopViewController() -> UIViewController? {
      guard var topViewController = UIApplication.shared.keyWindow?.rootViewController else {
        return nil
      }

      while let nextViewController = topViewController.presentedViewController {
        topViewController = nextViewController
      }

      return topViewController
    }
}

@available(iOS 11.2, *)
extension RNCardScan: SimpleScanDelegate {
    // MARK: -SimpleScanDelegate implementation
    func userDidCancelSimple(_ scanViewController: SimpleScanViewController) {
        if let topViewController = getTopViewController() {
            topViewController.dismiss(animated: true, completion: nil)
        }

        if let resolve = resolve {
            resolve([
                "action": "canceled",
                "canceledReason": "user_canceled"
            ])
        }
    }

    func userDidScanCardSimple(_ scanViewController: SimpleScanViewController, creditCard: CreditCard) {
        if let topViewController = getTopViewController() {
            topViewController.dismiss(animated: true, completion: nil)
        }

        var resolvePayload: [String: Any] = [:]
        resolvePayload["action"] = "scanned"
        resolvePayload["payload"] = {
            var payload: [String: Any] = [:]
            payload["number"] = creditCard.number
            payload["cardholderName"] = creditCard.name
            payload["expiryMonth"] = creditCard.expiryMonth
            payload["expiryYear"] = creditCard.expiryYear
            return payload
        }()

        if let resolve = self.resolve {
            resolve(resolvePayload)
        }
    }
}
